runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -0,0 +1,655 @@
+	U_8 returnType = 0;
+	bool returnsObject = false;
+	J9VMEntryLocalStorage newELS = {0};
+	J9VMThread *currentThread = NULL;
+	bool isCurThrdAllocated = false;
+	bool throwOOM = false;
+	J9Method* thrLiterals = NULL;
+	UDATA returnStorage = 0;
+	j9object_t sessionOrScopeObject = NULL;
+
+	/* Determine whether to use the current thread or create a new one
+	 * when there is no java thread attached to the native thread
+	 * created directly in native.
+	 */
+	currentThread = getCurrentThread(data, &isCurThrdAllocated);
+	if (NULL == currentThread) {
+		/* The OOM exception set in getCurrentThread() will be thrown in the interpreter
+		 * after returning from the native function in downcall.
+		 */
+		goto doneAndExit;
+	}
+
+	returnType = getReturnTypeFromMetaData(data);
+	returnsObject = ((J9NtcPointer == returnType) || (J9NtcStruct == returnType)) ? true : false;
+
+	VM_VMAccess::inlineEnterVMFromJNI(currentThread);
+	if (buildCallInStackFrameHelper(currentThread, &newELS, returnsObject)) {
+
+		/* Save the current executing method and restore later after the memory allocation
+		 * for pointer/struct in the argument list.
+		 */
+		thrLiterals = currentThread->literals;
+
+		/* The argument list of the upcall method handle on the stack includes the target method handle,
+		 * the method arguments and the appendix which is set via MethodHandleResolver.upcallLinkCallerMethod().
+		 *
+		 * Note: push the target method handle on the special frame so as to avoid updating the address
+		 * on the stack by GC when allocating memory for the next pointer/struct of the argument list.
+		 */
+		j9object_t calleeHandle = J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_CALLEEMH(currentThread, J9_JNI_UNWRAP_REFERENCE(data->mhMetaData));
+		PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, calleeHandle);
+
+		for (I_32 argIndex = 0; argIndex < paramCount; argIndex++) {
+			U_8 argSigType = sigArray[argIndex].type & J9_FFI_UPCALL_SIG_TYPE_MASK;
+
+			switch (argSigType) {
+			case J9_FFI_UPCALL_SIG_TYPE_CHAR:  /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_SHORT: /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_INT32: /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+			{
+				/* Convert the argument value to 64 bits prior to the 32-bit conversion to get the actual value
+				 * in the case of boolean/byte/char/short/int regardless of the endianness on platforms.
+				 */
+				I_64 argValue = *(I_64*)vmFuncs->getArgPointer(nativeSig, argsListPointer, argIndex);
+#if !defined(J9VM_ENV_LITTLE_ENDIAN)
+				/* Right shift the 64-bit float argument by 4 bytes(32 bits) given the actual value
+				 * is placed on the higher 4 bytes on the Big-Endian(BE) platforms.
+				 */
+				if (J9_FFI_UPCALL_SIG_TYPE_FLOAT == argSigType) {
+					argValue = argValue >> J9_FFI_UPCALL_SIG_TYPE_32_BIT;
+				}
+#endif /* J9VM_ENV_LITTLE_ENDIAN */
+				*(I_32*)--currentThread->sp = (I_32)argValue;
```

Comment:
This does not seem to work as expected on AArch64 macOS.
CHAR and SHORT are stored at 1-byte and 2-byte boundaries in the native stack on AArch64 macOS.
Reading the argument value from the native stack as I_64 and converting it to I_32 can pass wrong value to the Java side for small types.

Reply:
@knn-k, I just updated the code against your proposal on macOS/AArch64. Please double-check to see whether it works well as expected.

Reply:
It works with Jep424Tests_testLinkerFfi_UpCall on AArch64 macOS in my local environment.
```
===============================================
    Jep424Tests_testLinkerFfi_UpCall
    Tests run: 207, Failures: 0, Skips: 0
===============================================
```